### ISSUE TITLE
Simplify codebase ideas

### ISSUE THREAD

**Issue by cellog** [intljusticemission/react-big-calendar#95]
As initially referenced in #93 I would like to take a stab at simplifying the codebase.

Initial ideas:

1. pull out internationalization into a separate package.  react-big-calendar-moment, react-big-calendar-globalize
2. remove state, use props for everything unless there is literally no other way to do it
3. make all components pure
4. reduce file size to single page in an IDE for every file, split up large components

And from a development perspective:

1. unit test everything using mocha/enzyme/sinon and use jsdom in wallaby for instant test failure notification
2. remove tangled dependencies, create higher order components to do that, but keep the display logic clean and simple, and dumb.
3. use react-storybook for visual testing of components (https://github.com/kadirahq/react-storybook), this is by far the easiest way.  I usually tweak the component in there, then write the unit test both to verify invisible things and as a regression catcher.
4. move to es6 completely?  Let the user transpile as necessary, so no more lib/

Basically, I'll do these things for my own use, gradually over time, but I would like it to be useful to you also.

What do you think?  Anything strike you as a different direction for your priorities?

**Comment by jquense** [comment]
thanks for taking the time to write these, up! Here is some initial off the cuff feedback:

1. I get the appeal of splitting I18n stuff out into separate packages, and I've tried that approach before in other components. In my experience splitting them out, doesn't gain you anything, while increases the maintenance burden of the library and makes more work for a user. In this case the localization stuff is _required_ for the component to work at all, I'd like to not ship a component that requires the user to explicitly install another dep just to get  to "hello world". For folks concerned about extra bytes you can import directly from lib only the bits you want.

2. Do you have some specific instances in mind where you feel the Component state is unwieldy? AS it stands I think there is fairly minimal amount of state in components.

3. Pure components in this case may not possible as that would require that the user guarantee that their events are immutable, which I am not sure is a feasible requirement for everyone. I am always down to improve performance, I'd be great if there were specific identifiable "hot" areas that could use tuning before we try an make everything pure, which may not give much of a benefit for the cost.

4. Glad to have components split up, however, I wouldn't use "fits on a screen" as a heuristic for judging size, components that are small for the sake of being small doesn't serve development

##### Development

1. 👍  to unit tests we desperately need them (shame on me), my preferred stack would be karma, mocha, sinon, jsdom, and [teaspoon](https://github.com/jquense/teaspoon) since thats what we use in other projects and internally.

2. Can you flesh this one out? What do you mean by tangled components and what kinds of HoC's do you have in mind

3. that sounds cool, never used it before :)

4. Strong disagree on this one. shipping uncompiled code to a user is an anti-pattern. A user should not have to understand how to build from source to use the component. If folks want source code they can clone from github, but I think NPM distributions should be compiled. also as an FYI `lib` is checked in so its easy to install using npm form github without needing an install hook (its a weird thing i do that few others do :P)

Thanks again for wanting to help out its appreciated!

**Comment by cellog** [comment]
I forgot to put in my rationale for splitting up locale stuff!  Sorry.  Quickly:

there are 2 issues with dates to solve:

1) manipulation of date data to calculate steps, dates, and so on
2) representing this in text form to the user

#1 should happen no matter what internally.
#2 should by default be very simple, and be extensible for those who need i18n

Thus, the current locale stuff should be split into 2 pieces, in my opinion.  Most users would not need the localization, but for those who do, running npm i -S react-big-calendar-i18n-globalize and adding 1 line of code to use it would not be a huge burden.

This way, it works out of the box.  I will split my replies to the other stuff to the next comment

**Comment by jquense** [comment]
Slight nit, everyone needs localization, I think you mean that most folks are from the US or is fine with US date formats (which may or may not be true), first days of the week etc. You also may be underestimating how complicated, even simple, date formatting is :P, I've been down that rabbit hole and eventually you end up rewriting Moment.js. Also the date stuff doesn't always fall neatly down the line you've drawn (tho it mostly does). How one does date calculation is inherently culturally dependent, take for instance the first day of the week, without knowing that its impossible to layout the calendar without just picking an arbitrary day.

 The localizers are intentionally pluggable to allow for more lightweight approaches for folks that want them. It may be more productive to provide a "simple-date-localizer" in addition to the moment and globalize ones. Ultimately though we don't want to be US-centric by default, its unnecessary and doesn't serve the wide range of users the library has.

Localization and dates are _all_ about the edge cases, and in my experience rather then considering everything an edge-case its more productive to embrace the ugliness of dates and use robust solutions instead of optimizing for byte-size.

**Comment by cellog** [comment]
all of the suggestions not related to testing/storybook are inspired by [redux](http://redux.js.org/docs/introduction/Motivation.html)

Basically, it is currently very hard to take react-big-calendar and implement some of the action recording and undo/redo that redux design principles allow.  Swapping out components or customizing easily is also more complex than I think it could be.

I think a good first step is for me to start some refactoring using storybook.  I'll hold off on tests until there is a proof of concept.  I've started a new branch in my fork, and within the branch I'm doing new development in the src/components directory, so there will be no conflict if you do decide to merge with the existing work.

I like teaspoon a lot, actually.  Probably going to use it in my own stuff, as enzyme/jsdom is a lot clunkier :)

**Comment by cellog** [comment]
I plan to preserve the localization stuff you have here, it is very good.  I just want to try to streamline it, not make things US-centric.  If I fail, you'll be the first to know :).  Mainly I want to make sure my priorities are not wildly different from yours.  If you're already planning a refactor, for instance, that goes a different direction.

**Comment by jquense** [comment]
I'm happy to see what you can come up with! I would add a comment about how I think RBC and its feature set. There is a philosophical difference between something like Redux and big-calendar. Redux is a low level library that is intentionally light, terse, and unopinionated. whereas RBC is a large, complex component that is a lot more "batteries included". 

That spectrum is about tradeoffs between ease of use, fully featured and configurable. We don't want to make common use-cases hard to implement in RBC but we also don't want users to have to implement everything on top of it, It should work for 80% of people out of the box, and allow the other 20% to (within reason) configure what they want. 

That being said I built this primarily because we need one internally, so it serves that use case above others, I'm happy to accept PR's and features for stuff we may not use, but only insofar as I'm not going to be spending all my maintaining them :) And insofar as folks are contributing thats not a problem!

**Comment by jquense** [comment]
Thanks again for taking the time to work on stuff, it's appreciated!

**Comment by cellog** [comment]
Here, for your viewing pleasure, is a sample of converting TimeGutter into a pure element:

https://github.com/cellog/react-big-calendar/tree/streamline/src/components

Basically, it satisfies everything from above.

- the react elements will always return the same markup for the same input
- all props are immutable, even the dates, and even "now"
- i18n is handled outside the component, allowing customization at the extreme
- storybook shows each possible state

With the code on master, it is much more complex to isolate and test the code that adds "rbc-now" to the class, because you have to write a test that creates a date range that surrounds today.

Also, I found a bug - it doesn't highlight the current time if it falls into a cell that has no label, so I fixed that.  Discovered it because of visually testing edge cases with storybook.

It's not all single page, but the components are shorter with this design, and there is really only 1 logical operation on each line, which makes it easier to mentally conceive of the entire component.

Does this make more sense?

**Comment by jquense** [comment]
Looks good, factoring out TimeSlice seems like a reasonable thing to do here. Nice about finding a bug! A few comments:

> the react elements will always return the same markup for the same input

You can do this with without rewriting the component as a Function Component; Personally I find the component signature harder to read and scan with the default props in arguments, and the arrow function, and the propTypes below the component. I wouldn't rewrite components as function components just for the sake of it, it doesn't in and of itself add anything to maintainability, performance of understandability of the code base. 

> all props are immutable, even the dates, and even "now"

this is a good component for this, because none of the props aren't particularly complicated.

> With the code on master, it is much more complex to isolate and test the code that adds "rbc-now" to the class, because you have to write a test that creates a date range that surrounds today.

Sure but you need to do that anyway because what needs to be tested is whether it adds the label for the correct dates, not whether it adds a label at all based on a boolean prop.

> i18n is handled outside the component, allowing customization at the extreme

This is a not quite true TimeGutter is now using Moment by default, which means that Moment is a required dependency of component even if the user doesn't use Moment (if they use globalize for instance). The way localization works now already completely isolates the i18n from the components themselves, thats the point to why its structured that way, in addition the localization strategies are already completely customizable, swappable, etc, and structured to allow users to only require the one they want without the added bulk of the others, its a finely honed pattern from my experience doing this for lots of libraries over the last few years.

**Comment by cellog** [comment]
A couple of points:

React docs explicitly recommend using functions because of "future optimization possibilities," so it sounds like this is the direction things are heading, but I understand where you're coming from.  There is also no possibility of attaching a ref to a stateless function, so it would need to be used only for those components that don't need state.

Fortunately, converting from a stateless function to a React es6 class is at most 30 seconds of work, I've done it a couple of times, so I don't see that as a showstopper either way.  Might even be scriptable :)

As for the i18n, since I haven't yet tackled that aspect of the code, there is no way for you to see the evil master plan, so I suppose I will have to say it here!

What I ultimately plan is to have the top-level elements pass in the formatter, and turn the default formatters into that, instead of using moment.  The moment code is just for rapid development.  Because the formatter is a prop, this is very easy to replace, and default values will be easy to replace as well.

For most of my components, I use () => null as a default value, but I wanted the storybook to work without twiddling.  Since I haven't worked my way through the codebase yet, I'm just not sure what shape the i18n will end up taking yet, so don't consider any of that stuff final.  I *completely* understand how important the i18n is, I realize my code gives a different impression.  I see that as a secondary challenge once the code is more modular that may require another refactor.  Right now, I don't understand the code well enough yet to make a definitive statement about what I think is the best way to pass the formatters into the components.

I'm also not ruling out your existing solution, to be completely transparent.  But using it requires importing 4 files and running some code, so I wanted to see if I could isolate that from the other things I'm working on.

I'm working on the TimeGrid next, that is quite a beast!

**Comment by jquense** [comment]
I appreciate that you are thinking about this, and glad to have more eyes on this, but I would really just leave the localization alone, believe me. I have spent a lot of time thinking about how to do this in a way the has the right set of tradeoffs and plays well with the existing i18n strategies. Passing a formatter down from the top is just like the current solution except that it can be blocked by SCU updates which is not optimal, and a user would have to provide one for each instance of the component, which is more work then the current setup.

I also don't want to have to refactor components to normal components when I need to attach state or need a ref. I admit that is a personal preference but I have to do that at work and its terrible for iterating on components. Right now there is just no need to optimize the components in that way, especially for future optimizations. The component is already very fast with even thousands of events in view, so I am not worried about that

**Comment by cellog** [comment]
all right, I'll use classes, no problem.  As I said, until the refactoring gets going, I won't have an opinion about i18n that is of any use, so my best answer is "ok" at this point :)

**Comment by jquense** [comment]
I should re-phrase that i bit :P you should feel free to use function components for things that aren't likely to need a ref and aren't very complicated/have a lot of props. the TimeSlice component is fine as one! Things that hold a bunch of logic tend to need refactoring often enough that its a pain to switch.

Thanks again.

**Comment by jquense** [comment]
O and if you are using the class syntax feel free to you static property initializers so propTypes can go up top:

```js
class Foo extends React.Component {
  static propTypes = {
     foo: whatever,
  }
}
```
```

**Comment by cellog** [comment]
ah cool, I just moved some stuff into static.

It occurred to me that you may understand better my motivation for these changes if you had a use case.

Basically, I am looking to replicate the functionality of simplybook.me and google calendar.  This react component does the best google calendar mimic I've ever seen, but when I try to tweak it to do what simplybook.me can do, I can't.  These changes would make it flexible enough to render both.

For example, I just refactored the TimeGutter component.  Instead of using a test for evenness to determine whether to print a label or not, and iterating over all the blank spaces, each collection of time slices is abstracted into a group.  The label is printed at the top of the group, and then lines are printed.  With this change, a group can have more than 2 lines (up to 5, is what I've implemented).  The TimeGutter takes a new argument "slices" and passes that into the TimeSliceGroup.  It also uses it to multiply the step by the number of slices so the label is accurate.

This change would have been a lot of work with the old design, but was simple with the new.  This kind of flexibility is what I'm hoping to bring to the whole thingy.  Thanks for bearing with me!  I realize this is a mission-critical component for your work, and don't want to hurt that at all, so again if my changes diverge too far, we can either collaborate to tweak them later before merge, or I can fork, either way is fine for me, but I appreciate the chance to experiment very much.

**Comment by cellog** [comment]
Hi again,

This is just an update.  Progress is at https://github.com/cellog/react-big-calendar/tree/streamline/src

I had spoken about higher-order components but didn't have anything specific to show.  Now I do.

I expect to use higher-order components for selection, i18n formatting, and basic app state (which view is being used, and so on.

This update is about selection, which I have nearly finished implementing.  At this stage, there are only storybook stories, no automatic tests, that will be the next step, as converting stories into tests is an easy way to develop.  Storybook is basically manual testing.

So what I noticed was that implementing selection in the existing project is complex, and requires a lot of boilerplate in each class that uses it.  So, I started out by re-thinking what selection is.  Basically, in the calendar, we want to be able to capture sequential selection of dates or times.  Dates and times are represented as boxes within a larger box.  We don't want to be able to select dissimilar types (dates and times, for instance).

Abstracting this into concepts, we want to have discrete containers of selectable things.  If a selection begins within a container, it should be limited to items within that container.  We want to be able to mark individual boxes as being selected, and also do more complex operations involving all of the boxes.

Abstracting further, we are selecting BOTH the boxes, and the values that they represent, and want to be able to work with both.

So, to implement this, I created a selection container creator (the box around things we want to be selectable), and a selectable box creator (the selectable items).  These are found in:

containers/Selection.jsx
containers/Selectable.jsx

Both contain functions that return a React Component.  To create a selection box you run:

`
class Thing extends Component { render() {...} }
const SelectableThing = Selection(Thing)
`

Then you can use SelectableThing anywhere you'd use a Thing.  To enable selection, you pass selectable.  There are a couple varieties of ways to implement selection, and I can go into those, but that is just implementation details.

`<SelectableThing selectable constantSelect />`

causes the behavior "selectable" does in the existing code.

To implement a selectable component, one has to tell the system how to identify the component uniquely (it's key), and what value it represents.  For example, a slice of time in the day view represents a start time.  End time is implied by the `step` prop and in my implementation the `slice` prop (which is 2 by default).

`
class TimeSlice extends Component { render() {...} }
const SelectableTimeSlice = Selectable(TimeSlice, {
 key: (props) => props.value.toString(),
 value: (props) => props.value
})
`

substitute SelectableTimeSlice anywhere you'd use TimeSlice.  You can see this in the code.

This removes hundreds of lines of code.

Then, to implement the larger stuff, you just pass in a onSelectSlot() callback to the Selection component. It is passed a map of key => selected object, a map of key => selected values object, and arrays of the objects, array of the values

This can then be used to implement the instant selection box or whatever else makes sense.

Any feedback at this stage?

**Comment by jquense** [comment]
Hi, I haven't had a chance to try anything, but not sure I agree that what you've done for selection here is simpler. There is a lot more code and levels of indirection now whereas before it was straightforward and the logic was up front. 

I agree that you've got a more generalized way to handle selection but that's overkill for the library which doesn't need HoCs or easily extendable selection paradigms like react dnd or something. 

I also want to add a friendly warning here, it's perfectly fine to play around with the stuff but there is almost no chance that I'll be able to merge a complete rewrite like this for a few reasons. The changes are too dramatic and without tests documenting the original behavior there is no way to have confidence that it doesn't break something for our production uses. Also it's a ton of code that I now have not written that I have to learn and ultimately maintain, which doesn't gain us much as the original code doesn't particularly have any problems for most uses.

I do thinks it's great you are able to experiment with the code base and explore new ways to do stuff, but remember that we have to weigh practical concerns before code organizational ones

**Comment by jquense** [comment]
Looking a bit more at the selection code, I do think that's worth pulling out into a separate library (I'd use it for other stuff). I did notice that the way time selection is handled is different then it is now. You've got the actual time slice blocks being selected vs an overlaid  mask that exists now. It does unify the selection strategy and simplify the month selection use case but it's also significantly changing the UX of the time selection use case. The difference in selection paradigm between the two cases was one of the reasons why the code was less unified on that front.

**Comment by jquense** [comment]
Sorry noting that you _do_ maintain the same visual overlay, but still it feels like it's stretching the API in an uncomfortable way here.

**Comment by jquense** [comment]
hey @cellog, On a more productive note, I think there are some stuff here useful and helpful things that would be great to incorporate into the code base. And you are clearly thinking about about this and working hard on it. I'd like to try and make sure that's productive work for you (and me), so I've been thinking of a few things I think that we can do to help make this process a bit more productive on that front though, so here are my suggestions:

1. It's hard to review code just sitting in a fork, instead why not PR something, it doesn't need to be finished but in a PR I can leave comments on the actual code instead of this soon to be giant issue.

2. Lets try and do work in chunks so that its easier to review and make sure something doesn't break. Incremental work, while slower is a lot easier to merge.

3. Pick one use-case at a time, that you'd like to use the calendar for but can't because of the current API, setup, etc. Lets discuss it before you rewrite the whole package and see if its a good fit for this, and the best place to start to make that happen. Grounding a change, or set of changing in a use case, can help me understand why its worth doing, and make sure the work is productive towards a specific goal, and not just work for its own sake (we all fall into that as programmers!)

**Comment by cellog** [comment]
You're so patient, I appreciate the steps you took to go from the initial "wtf is this" to "ok here's how I am comfortable working with this" :)  I've done open source long enough to know this is a very unusual level of patience, so thank you.  I'm going to test that patience in this comment, it's a loooong one, so hopefully you aren't twitching and frothing at the mouth by the end!

Let's start with some use cases I have immediate need for, and I think it may make more sense.

I'm working at a music school in the string area at the University of Nebraska-Lincoln.  We have a need to handle sign-ups for rare events.  As a specific case, at the end of every semester, there are 2 days in which string students perform a short subset of the work they have done that semester for the faculty.  There are different categories of student, and their jury length varies from 10 to 20 minutes based on the category.  For example, non-majors and music minors play a 10 minute jury.  music majors and graduate students play a 15 minute jury.  Students at the end of their sophomore year have to play a special jury whose passing grade is required to continue as a music major, and that is 20 minutes long.  So, when students choose a jury time, it needs to be exactly that length, and can't be an arbitrary length, the way most event scheduling on a calendar needs to be.  At the same time, we need to block off specific times that the students can sign up for their juries (like Mon/Tue 9-10, 10:15-12, 1-3, 3:15-5, for instance), and those vary from semester to semester, depending on the number of students playing a jury.  In addition, individual faculty members often have conflicts with a subset of the jury times, and ask their students to sign up for times that they will be present for the jury.

What I am writing is an app that students and faculty can log into that links them with each other, so that we have enough data to be able to present to a student the exact times they can sign up for (all jury times excluding their teacher's unavailability) and limits them to signing up for the correct length of jury for the semester.  I want them to be able to click-drag and have the jury time slide based on whether the times are actually available and long enough to accomodate the jury length.  I also want to have a separate view for the faculty where 2 things can happen

1. the head of the string area can choose dates for the jury, and then click-drag on the empty slots to choose the times in which juries will happen
2. the individual faculty can log in, and click-drag on times that they will be unavailable, choosing only from those times.

and then a UI concern:

3) because the juries are organized in 10-20 minute chunks, we need 5-minute time slots.  With that many time slots, having a thinner time slot would make it possible to view the entire day on a single screen

Then, finally, I need a view where anyone can view the current juries in a straightforward calendar and then print a list of juries for the faculty to use in the jury room (this last part is separate from the calendar, it's just formatting the event data into a table, so doesn't involve the calendar, but I am including it for completeness).

I took a look at the current design, and implementing the student view would require rewriting the selection code inside TimeGrid.jsx, since it is tightly coupled to the current paradigm of selecting adjacent slots, and the code to display the selected slot is inside TimeGrid.  Extending TimeGrid.jsx might be a solution, but then it requires extending all the code that uses this, and duplicating a lot of the code in order to switch out a line here and there.  In addition, because DaySlot has its own selection code, it also would need to be extended in the same manner.  TimeGutter hard-codes in 2 slots per time label, as you know, and pulling out TimeSlice and abstracting further to TimeSliceGroup allows changing the number of slots per step to 2-5 (with some change to the css, of course, to make the segments thinner).

In addition, because the display and capture of selection information is inside TimeGrid.jsx, it makes testing the selection display much more difficult because with the same props, the component can display differently, based on uncontrolled input from mouse events.  Short version: it's not a pure component.

So that is the set of challenges as I see them.  I would like to separate them out into chunks for you.  Here is the discrete chunks as I see them:

1. easiest chunk: making it possible to display 2-5 slices per step.  This is low-level and not hard at all
2. hard: making selection generic enough that it can be used for all the use cases above

for number 2, I'm pretty sure it requires re-factor rather than tweaks to the existing design, and that's why I wrote the stuff I wrote.  My primary design goals were to avoid any unnecessary duplication, try to make each component only do the thing that its specific visual space on the calendar implies (blank space should be selectable to allow choosing that time slot, a column should represent the times on a day, the size of the blank space should represent how many minutes are within it, and so on) and to make it possible to understand what a component is with a glance, if possible.  Or, put simply, to make components focused and easy to maintain.

Along those lines, I noticed that the TimeGutter was serving the same purpose as the overlay in the DaySlot.  Namely, splitting up time within a day into discrete chunks of a specific length (10 minutes by default).  The only difference visually is whether a label is printed (1:05pm for instance) or not.  So, one of the first tasks I saw was to unify that codebase.  The only substantial difference logically between the two is that the gutter on the left is only a legend, whereas the gutters within day slots represent actual time slots, and should be selectable.  So, I decided to extract TimeSlice, and then TimeSliceGroup in order to make it simpler to represent a series of slices that have a label.

Then, I extracted the selection code from all the UI components, and tried to capture the essence of what was being selected.  In the original code, slice ID was the sequence number of the for loop that generated the slices.  I thought about this for a while, and realized that the sequence number was an abstraction for the actual value of the slice, which is the times it represents.  I chose to store that value as the start time of the slice, since the length of the slice is determined outside the slice by the higher code anyways, but it could have just as easily been a time range for a different application.  Also, the label is generated from the start time, so that is the primary piece of data needed for the UI.

With these changes, it then becomes simple to re-use TimeGutter.  The DaySlotTimeGutter (which I called DayCells because I'm stupid and wasn't thinking, but will rename shortly) is the TimeGutter, but with a selectable timeslice popped in.  That's it!  No other code change necessary.  I implemented it as 2 components, but it could be simplified to passing in the component to use for TimeSlice, and then using TimeGutter directly.  I didn't think of this when designing it, only just now when typing this, but it seems the better option to me, because then selection can be implemented with a single line, something like:

`<TimeSliceGroup timeslice={this.props.selectable ? SelectableTimeSlice : TimeSlice} />`

The other major change I made was, as you noted, I linked the data directly to the UI in terms of the time being given to each visible slot.  This was a change that I also think is important.  If the data is linked to the thing being selected, then we can easily collect the values that the user is selecting, which is how users think about selection.  Then we can pass that to the overseeing onSelectSlot callback, as well as the actual DOM nodes that encapsulate the data being selected.  Ultimately, this will make it much easier to test the code that creates the selection overlay to confirm that the correct calculations are being passed in, although there is still work to be done (findDOMNode needs to be mockable to inject values into the calculation for getBoundsForNode, for instance).

There are a few major issues that I know about

1. if you start the storybook (npm run storybook) and browse to http://localhost:9002, then click on the DaySlot story, and click-drag to make a selection, you'll see that the mouseup doesn't always register as ending the selection, and then when doing mousemove, it will do all kinds of weird stuff including selecting a reverse time slot.  This is because I tried a slight variant on your existing selection code for mouse events, and it isn't working right :).  I will need to work with you to fix that, since you have many more years experience working with mouse events and the edge cases.

2. the code appears on the surface to be dramatically different from your code, even though the internals are really exactly the same design principles, and as you say, tests are needed for the old code before replacement can even be attempted.

3. i18n is not specifically handled yet, and is another use case to discuss.  Currently, all the i18n is placeholder stuff based on moment, and that cannot be merged into the codebase.  I am thinking of simply replacing the placeholder with the global formatter variable you have, and tackling that question in a later PR.  Because my immediate need is localized in Nebraska, and the i18n code is working for you and your current users, I'm not as concerned in the short term about this.  Long-term, I'd like to make the component pure (no global variables) for testing purposes as well as other reasons, but that is a separate question.

And lastly just to review, I took the steps that I did because I want to ensure that there is no chance that a pull request could become tangled up.  What I want to do is:

1. write a proof-of-concept for you to see (done)
2. get your desires/needs reflected in the code (working on it)
3. decide on what is acceptable as a design based on 2 (future/now)
4. write tests for the original code that will make dropping in the new code painless (near future)
5. write a pull request once these are done

For 4, I would also want to write storybook stories for the old code, so that you can eyeball the two and make sure that the new code looks like the old code as well.

My time frame for finishing all of this to make it useful for next semester is a deadline of November.  Jury signups won't happen before then.  If we finish sooner, great.  If not, I can still use simplybook.me until it's finished.

And (a la Steve Jobs) just one more thing:

If these things are ultimately separated out enough so that all the UI is pure, it will allow plugging redux directly into the calendar.  Then I can take advantage of redux-specific advantages, such as the ability to record and replay actions both to reproduce bugs and for unlimited undo/redo, all accomplishable with literally 1 line of code.  But most important: the design I'm envisioning would not force redux as a dependency upon users, redux is more of a thought framework than a code framework anyways, it's a way of organizing code that makes changing things simpler and more robust rather than a set of widget requirements like a framework.

If you're still reading this, thanks again, and congratulations, you survived one of my treatises disguised as a comment! :)

**Comment by jquense** [comment]
Ok, thanks for the explanations, from all that this is what I think makes the most sense from what I understand of your use case.

I think it makes sense to open up a PR that refactors TimeGrid into something that is more generic and customizable, by both length, and groupings. Right now the code is already fine for variable length (week, work weed, day, extra), it also handles min/max fine for the total length of of a DaySlot. If you adjust the `step` prop to `5` instead of the default `30` and I think you get most of what you want already. It sounds like a good idea to implement the TimeGutter in the same way DaySlot is, then all you need is a prop function that returns `true` for the slot you want labels on.

It doesn't seem necessary to add the concept of Groupings here, or subStep slices, when you could just increase/decrease the `step` size to your 5min increments, Selection will also just work then without any changes. If you do want to add a group, switching selection to calculate off of the sub slice would be easy as well but doesn't (as far as I can see) require a change of strategy with regards to selection, in fact the current approach is nicely extensible here. 

Ultimately though, selection is designed in terms of the _data_ first and tries as hard as it can to not have any opinion about underlying DOM nodes, no state in the DOM so to speak. Honestly tho that choice makes it really easy to just drop in for your situation. 

I would just isolate changes to that refactor and leave the rest of the stuff (i18n, pure components, etc) alone for now because it doesn't practically make much of difference. Agreed that some things are a bit tough to test, (like selection) but I don't think those things really need or benefit from unit tests anyway.

In terms of redux integration, I don't quite get what you mean there, we use RBC with redux in lots of places, and it works great, and with no problems already. The Calendar tracks almost zero internal state, and the state it does track is view state and doesn't belong in a Redux store anyway. All the business logic stuff, event editing, navigation, view selection, etc is all controllable via props already.

**Comment by cellog** [comment]
OK, I'll take a look at these changes.  Thanks!

**Comment by cellog** [comment]
Just a quick note:

I'm on vacation this week, so pace will slow.

Yesterday I figured out all the issues in selection, so now it works exactly like the original.

One issue I found in both selection implementations is if you wish to click-drag past the bottom of the screen, it isn't possible.  Google calendar slowly scrolls.  If I can figure out how to implement that, then the slices code becomes unnecessary, and I'll remove it.  I also wanted to implement an additive selection mode, where you can select by clicking, and it stays, then deselect by clicking.  This will make choosing busy/free times really easy.  I also wanted to experiment with adding support for touch events, and see if that is easy or not.

I will unit test selection.  It is worth it.  I already had a really tricky to detect bug which a unit test would have been able to see with no trouble.

I also have to admit I read your comment several times, and I am pretty sure we are on the same page, but I want to make sure :)

So what I am doing is taking the TimeGrid, splitting selection out into Selection/Selectable, making TimeGutter generic, with a simpler TimeSlice and TimeSliceGroup than what I have now (removing slices, sticking with steps), and making DaySlot/TimeGutter slightly different implementations of the same class.  And testing the jeans off of the original so I can detect regressions, then making PR as I go.  Being as small and modular in the changes as possible (i.e. lots of small PR).

I have one request, and let me know if you want me to open a separate issue for it: can you make sure the test setup is the way you want it?  I will plug in unit tests in any format you prefer, I'm comfortable and have tons of experience being super flexible with that.  That way, I can write tests that you will immediately find useful, and PR them.

**Comment by jquense** [comment]
Feel free to PR stuff right from the beginning, and just use it as a public work in Progress, That is also easier for me to add comments, and notes as you go along, and less likely to lose some of the institutional knowledge around why stuff is the way it is. 

again first just implement the stuff in TimeGrid to cover your slot needs, we can deal with additional features, like autoscroll and what after, where the added complexity is more clear and easier to evaluate.

**Comment by cellog** [comment]
ok, did the first step, #102

**Comment by cellog** [comment]
I have been thinking quite a bit about the source of the friction between my pull requests and your needs, and I have another perspective to throw at you, see what you think.

One of the things I've noticed, as I have been working on refactoring, is that there are a lot of props that are getting passed down.  In most of the subcomponents, many of the props like formats and so on that are just passed by the calendar are passed with {...props} and so not even documented with the propTypes map.  I made all of these explicit, especially the formats that are actually used, in each component.

Because the props all need to be passed down, and there are so many to keep track of and ensure they get passed to the children properly, there are 3 ways of handling this:

1. what you have done, which is to use ...props and trust that it will work.  The advantage is that it works with existing layout and is very simple. Disadvantage is that refactoring is difficult for anyone but the author, as it is not at all clear what each component does or where its props come from without studying the whole source.
2. specify global stuff in context instead, and use contextTypes in the children to pull in only the values needed.  This could be a good solution, but makes testing harder, as each test has to mock the parent.  Creating a mock that can be used for all tests could solve this.  Advantage is much shorter prop lists.  The disadvantage is that you ahve ot manually manage the context, and it makes it much harder to re-configure the calendar for unusual use cases.
3. wrapping all the stuff that would go in context into a single object, and passing that object as context would make things less brittle.  You could then pull out data as needed from the store.  Extending this idea further, it would be nice if the callbacks were also in that object, so you could call them directly from it, without having to update contextTypes for each component.  Once you get to this point, it's worth noting that this object is called a "store" in redux, and it would make a lot of sense to use this internally.  redux can manage the addition of new properties without interfering with the old, and it's very simple to test.

The third way allows subscribing to internal events from outside the calendar, so that you can subtly manipulate behavior without having to rewrite the whole thing with 1 change.  It would allow substituting a different behavior in events without having to pass in a complex prop list, etc. etc.

This is the mindset I'm approaching my refactoring from, and why I'd like to make the components as pure as possible.  This makes far more advanced stuff simple, and basic stuff simple too.
